### PR TITLE
cl_khr_icd 2.0.0 loader managed dispatch

### PR DESCRIPTION
**Pull request by Kerilk**
This implements the loader managed dispatching strategy discussed in https://github.com/KhronosGroup/OpenCL-Docs/issues/1003

### CODE REVIEW

File: loader/icd.c [hunk 1]
```diff
@@ -132,27 +144,77 @@ void khrIcdVendorAdd(const char *libraryName)
         char *suffix;
         size_t suffixSize;
 
-        // call clGetPlatformInfo on the returned platform to get the suffix
-        if (!platforms[i])
+        // skip NULL platforms and non dispatchable platforms
+        if (!platforms[i] || !platforms[i]->dispatch)
```
Comment (s-barannikov) [inline loader/icd.c:148]: How can dispatch be NULL?

File: include/cl_khr_icd2.h [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+#define CL_LAYER_API_VERSION_200 200
+
+typedef cl_int CL_API_CALL
+clDeinitLayer_t(void);
+
+typedef clDeinitLayer_t *pfn_clDeinitLayer;
+#endif //defined(CL_ENABLE_LAYERS) && !defined(CL_LAYER_API_VERSION_200)
+
+#if !defined(CL_ICD2_TAG_KHR)
+#define CL_ICD2_TAG_KHR ((size_t)0x4F50454E434C3331ULL)
```
Comment (s-barannikov) [inline include/cl_khr_icd2.h:15]: This doesn't fit size_t on 32-bit platforms, potentially causing compiler warnings.

Comment (Kerilk) [reply include/cl_khr_icd2.h:15]: Indeed the expected tag on 32 bit platforms would be `0x434C3331`. I will update the PR with a detection for 32 bit platforms.

File: loader/icd.c (hunk 1 above)
Comment (Kerilk) [reply loader/icd.c:148]: I just extended the original check, to ensure that the returned platform is valid. The idea is not to register platform that obviously do not match the OpenCL icd constraints.

File: include/cl_khr_icd2.h (hunk 2 above)
Comment (Kerilk) [reply include/cl_khr_icd2.h:15]: Found the warning on msvc:
`warning C4310: cast truncates constant value`

Comment (Kerilk) [reply include/cl_khr_icd2.h]: Replaced with:
```C
#if !defined(CL_ICD2_TAG_KHR)
#if INTPTR_MAX == INT32_MAX
#define CL_ICD2_TAG_KHR ((intptr_t)0x434C3331)
#else
#define CL_ICD2_TAG_KHR ((intptr_t)0x4F50454E434C3331)
#endif
```
Hopefully portable enough.

**Pull request closed without merging**